cargo-concordium/src/context.rs

```diff
@@ -0,0 +1,122 @@
+use anyhow::anyhow;
+use concordium_contracts_common::{
+    AccountAddress, Address, Amount, ContractAddress, OwnedPolicy, SlotTime,
+};
+use wasm_chain_integration::{ExecResult, HasChainMetadata, HasInitContext, HasReceiveContext};
+
+/// A chain metadata with an optional field.
+/// Used when simulating contracts to allow the user to only specify the
+/// necessary context fields.
+#[derive(serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct ChainMetadataOpt {
+    slot_time: Option<SlotTime>,
+}
+
+impl HasChainMetadata for ChainMetadataOpt {
+    fn slot_time(&self) -> ExecResult<SlotTime> { unwrap_ctx_field(self.slot_time, "slot_time") }
+}
+
+/// An init context with optional fields.
+/// Used when simulating contracts to allow the user to only specify the
+/// necessary context fields.
+#[derive(serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct InitContextOpt<Policies = Vec<OwnedPolicy>> {
+    metadata:        Option<ChainMetadataOpt>,
+    init_origin:     Option<AccountAddress>,
+    sender_policies: Option<Policies>,
+}
+
+impl InitContextOpt {
+    pub fn new() -> Self {
+        Self {
+            metadata:        None,
+            init_origin:     None,
+            sender_policies: None,
+        }
+    }
+}
+
+impl HasInitContext for InitContextOpt {
+    type MetadataType = ChainMetadataOpt;
+
+    fn metadata(&self) -> ExecResult<&Self::MetadataType> {
+        unwrap_ctx_field(self.metadata.as_ref(), "metadata")
+    }
+
+    fn init_origin(&self) -> ExecResult<AccountAddress> {
+        unwrap_ctx_field(self.init_origin, "init_origin")
+    }
+
+    fn sender_policies(&self) -> ExecResult<&Vec<OwnedPolicy>> {
+        unwrap_ctx_field(self.sender_policies.as_ref(), "sender_policies")
+    }
+}
+
+/// A receive context with optional fields.
+/// Used when simulating contracts to allow the user to only specify the
+/// necessary context fields.
+#[derive(serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct ReceiveContextOpt<Policies = Vec<OwnedPolicy>> {
+    metadata:                Option<ChainMetadataOpt>,
+    invoker:                 Option<AccountAddress>,
+    self_address:            Option<ContractAddress>,
+    pub(crate) self_balance: Option<Amount>,
```

Comment:
Why not make all the fields pub(crate), or alternatively, why does this one need to be public?

Reply:
This particular field is changed if you supply the `--balance` parameter during contract simulation. I've added a comment about this.